File: 1082D.MaximumDiameterGraph/maximundiametergraph.py
a = []  # degree maximo para cada vertice 
n = 0  # la cantidad de vertices
Ones = []  # array de index con veritces con degree 1 


def solution ():
    s = 0 
    for item in a :  # esto es para contar la suma de todos los degree  
        s = s + item 
    
    if s < 2*(n-1) :# si no hay la cantidad minima requerida para armar un grafo conexo 
                    # entonces imprimo NO
        print("NO")
        return
    
    for i in range(n): # guardo los indices de los vertices cuyo degree es 1 
        if a[i] == 1 :
            a[i] = 0 
            Ones.append(i)

    t = len(Ones)
    d = (n-t) - 1 + min(2,t)# el diametro va a ser el tamano del arbol que se pueda 
                            #formar con los conjunto de vertices con degree mayor que 1 
    print("YES " + str(d) + "\n" + str(n-1) )

    l = -1 
    if len(Ones) != 0 :  # l lo pongo como el ultimo vertice en la secuencia con degee 1
        l = Ones[len(Ones)-1]
        Ones.remove(l)

    for i in range (n):  # armo el arbol que estoy buscando 
        if a[i] > 1 :
            if l !=-1:
                a[l] = a[l] -1 
                a[i] = a[i] -1 
                print (str (l+1) + " " + str(i+1))
            l=i
    
    i = n-1   # anado aristas sin armar ciclos para que el diemetro no disminuya 
    while i >=0 :
        while len(Ones) > 0 and a[i] > 0 :
            a[i] = a[i] -1 
            u = Ones[len(Ones)-1]
            print(str(i+1) + " " + str (u +1) )
            Ones.remove(u)
        i = i -1 



if __name__ == "__main__":
    line = int(input())
    n = int(line)

    line = str(input()).split()
    for i in range (n):
        a.append(int(line[i])) 

    solution()  # esto imprime la solucion del problema 
    
        

File: 1082D.MaximumDiameterGraph/maximundiametergraph.cpp
#include <bits/stdc++.h>

using namespace std ; 

int n , a[1000] ;  /*el numero de vertices de grafo y el array que me dice el 
                    mayor degree que puede tener cada vertices*/
vector <int> Ones ; /* para saber cuales vertices tienen degree 1 */

void solution () 
{
    int i,  t  , d , l ,s =0  ; 
    for ( i =0 ; i < n ; i++) /*sumo todos los degree del array a */ 
    {
        s = s + a[i]; 
    }

    if (s < 2*(n-1)) /* si no se puede armar un grafo conexo entonces imprimo NO*/ 
    {
        cout << "NO" << endl ; 
        return; 
    }

    for (i =0 ; i< n ; i++)
    {
        if (a[i] == 1 )
        {
            a[i] = 0; 
            Ones.push_back(i);
        }
    }


    t = Ones.size(); 
    d = (n-t) -1 + min(2,t) ; /* el diametro que va a tener el grafo para armar */ 

    cout << "YES" << " " << d << endl << n-1 << endl ; 

    l = -1 ; 
    if (!Ones.empty())
    {
        l = Ones.back(); 
        Ones.pop_back(); 
    }

    for (i = 0 ; i< n ; i++)
    {
        if (a[i] > 1 ) 
        {
            if (l !=-1)
            {
                a[l] = a[l] - 1;
                a[i] = a[i] - 1;
                cout << l +1 << " " << i+1 << endl ;   
            }
            l = i ; 
        }
    }

    for (i = n-1 ; i >=0 ; --i )
    {
        while (!Ones.empty() && a[i] > 0 ) 
        {
            a[i] = a[i] - 1 ; 
            cout << i+1 << " " << Ones.back() + 1 << endl ; 
            Ones.pop_back();   
        }
        
    }    
}

int main () 
{
    cin >> n ;
    int i; 
    for (i=0 ; i < n ; i++)
    {
        cin >> a[i] ;  
    } 
    solution() ;
}